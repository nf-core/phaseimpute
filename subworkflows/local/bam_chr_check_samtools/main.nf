include { BAMCHREXTRACT as BAMCHRBFR  } from '../../../modules/local/bamchrextract'
include { GAWK                        } from '../../../modules/nf-core/gawk'
include { BAM_CHR_RENAME_SAMTOOLS     } from '../bam_chr_rename_samtools'
include { BAMCHREXTRACT as BAMCHRAFT  } from '../../../modules/local/bamchrextract'

workflow BAM_CHR_CHECK_SAMTOOLS {
    take:
    ch_bam          // channel: [ [id], bam, bai ]
    ch_fasta        // channel: [ [genome], fasta, fai ]

    main:

    ch_versions = Channel.empty()

    // Get contig names from the BAM
    BAMCHRBFR(ch_bam.map{ meta, bam, index -> [meta, bam] })
    ch_versions = ch_versions.mix(BAMCHRBFR.out.versions)

    // Check if the contig names are the same as the reference
    chr_disjoint = check_chr(BAMCHRBFR.out.chr, ch_bam, ch_fasta)

    if (params.rename_chr == true) {
        // Generate the chromosome renaming command based on the fasta file
        GAWK(ch_fasta.map{ metaG, fasta, fai -> [metaG, fai] }, [])
        ch_versions = ch_versions.mix(GAWK.out.versions)

        chr_prefix = GAWK.out.output
            .splitCsv()
            .map{ it[1][0] }

        // Generate the chromosome renaming file
        BAM_CHR_RENAME_SAMTOOLS(
            chr_disjoint.to_rename.map{meta, bam, index, nb -> [meta, bam, index]},
            chr_prefix
        )
        ch_versions = ch_versions.mix(BAM_CHR_RENAME_SAMTOOLS.out.versions)

        // Check if modification has solved the problem
        BAMCHRAFT(BAM_CHR_RENAME_SAMTOOLS.out.bam_renamed.map{ meta, bam, csi -> [meta, bam] })
        ch_versions = ch_versions.mix(BAMCHRAFT.out.versions)

        chr_disjoint_after = check_chr(BAMCHRAFT.out.chr, BAM_CHR_RENAME_SAMTOOLS.out.bam_renamed, ch_fasta)
        chr_disjoint_after.to_rename.view()
        chr_disjoint_after.to_rename.map{
            error 'Even after renaming errors are still present. Please check that contigs name in bam and fasta file are equivalent.'
        }
        ch_bam_renamed = BAM_CHR_RENAME_SAMTOOLS.out.bam_renamed

    } else {
        chr_disjoint.to_rename.map {
            error 'Some contig names in the BAM do not match the reference genome. Please set `rename_chr` to `true` to rename the contigs.'
        }
        ch_bam_renamed = Channel.empty()
    }

    ch_bam_out = chr_disjoint.no_rename
        .map{meta, bam, csi, chr -> [meta, bam, csi]}
        .mix(ch_bam_renamed)

    emit:
    bam            = ch_bam_out            // [ [id], bam, csi ]
    versions       = ch_versions           // channel: [ versions.yml ]
}


def check_chr(ch_chr, ch_bam, ch_fasta){
    chr_checked = ch_chr
        .combine(ch_bam, by:0)
        .combine(ch_fasta)
        .map{metaI, chr, bam, csi, metaG, fasta, fai ->
            [
                metaI, bam, csi,
                chr.readLines()*.split(' ').collect{it[0]},
                fai.readLines()*.split('\t').collect{it[0]}
            ]
        }
        .map { meta, bam, csi, chr, fai ->
            [meta, bam, csi, (chr-fai).size()]
        }
        .branch{
            no_rename: it[3] == 0
            to_rename: it[3] > 0
        }
    return chr_checked
}
