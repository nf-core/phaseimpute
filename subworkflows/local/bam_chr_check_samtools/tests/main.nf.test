nextflow_workflow {

    name "Test Subworkflow BAM_CHR_CHECK_SAMTOOLS"
    script "../main.nf"

    workflow "BAM_CHR_CHECK_SAMTOOLS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_chr_check_samtools"
    tag "bam_chr_check_samtools"
    tag "subworkflows/bam_chr_rename_samtools"
    tag "bam_chr_rename_samtools"

    tag "gawk"

    test("No rename: bam chr + fasta chr") {
        config "./nextflow_rename.config"
        when {
            workflow {
                """
                fai_file = Channel.of('chr22\t10000\t7\t60\t61', 'chrM\t10000\t7\t60\t61').collectFile(name: 'chr22M.fai', newLine: true)
                input[0] = Channel.fromList([
                    [
                        [id: "test_paired"],
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true)
                    ],
                    [
                        [id: "test_MT"],
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam_bai'], checkIfExist:true)
                    ],
                ])
                input[1] = Channel.of([[id:"GRCh37chr"],[]])
                    .combine(fai_file)
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
    test("Rename: bam chr + fasta no chr") {
        config "./nextflow_rename.config"
        when {
            workflow {
                """
                fai_file = Channel.of('22\t10000\t7\t60\t61', 'M\t10000\t7\t60\t61').collectFile(name: '22_M.fai', newLine: true)
                input[0] = Channel.fromList([
                    [
                        [id: "test_paired"],
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true)
                    ],
                    [
                        [id: "test_MT"],
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam_bai'], checkIfExist:true)
                    ],
                ])
                input[1] = Channel.of([[id:"GRCh37"],[]])
                    .combine(fai_file)
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("Rename: bam no chr + fasta chr") {
        setup {
            run("BAM_CHR_RENAME_SAMTOOLS", alias: "PREPROCESS") {
                script "../../bam_chr_rename_samtools/main.nf"
                process {
                    """
                    input[0] = Channel.fromList([
                        [
                            [id: "test_paired"],
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true)
                        ],
                        [
                            [id: "test_MT"],
                            file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam'], checkIfExist:true),
                            file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam_bai'], checkIfExist:true)
                        ],
                    ])
                    input[1] = Channel.of("nochr")
                    """
                }
            }
        }
        config "./nextflow_rename.config"
        when {
            workflow {
                """
                fai_file = Channel.of('chr22\t10000\t7\t60\t61', 'chrM\t10000\t7\t60\t61').collectFile(name: 'chr22M.fai', newLine: true)
                input[0] = PREPROCESS.out.bam_renamed
                input[1] = Channel.of([[id:"GRCh37chr"],[]])
                    .combine(fai_file)
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
    /*
    test("Error : missing renaming params") {
        config "./nextflow.config"
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "test_paired"],
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true)
                    ],
                    [
                        [id: "test_MT"],
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam_bai'], checkIfExist:true)
                    ],
                ])
                input[1] = Channel.of([[id:"GRCh37"],[],
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/genome.fasta.fai",checkIfExist:true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.failed },
                { assert workflow.errorReport.contains("Some contig names in the VCF do not match the reference genome. Please set `rename_chr` to `true` to rename the contigs.")}
            )
        }
    }
    test("Error : still difference after renaming"){
        config "./nextflow_rename.config"
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    input[0] = Channel.fromList([
                    [
                        [id: "test_paired"],
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true)
                    ],
                    [
                        [id: "test_MT"],
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam_bai'], checkIfExist:true)
                    ],
                ]) // Error due to multiple contigs name in header not present in fasta file
                input[1] = Channel.of([
                    [id:"GRCh37"],
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/genome.fasta",checkIfExist:true),
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/genome.fasta.fai",checkIfExist:true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.failed },
                { assert workflow.errorReport.contains("Even after renaming errors are still present. Please check that contigs name in vcf and fasta file are equivalent.")}
            )
        }
    }
    */
}
