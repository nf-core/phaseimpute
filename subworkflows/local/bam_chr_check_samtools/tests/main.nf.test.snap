{
    "Rename: bam chr + fasta no chr": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "test_MT.bam:md5,33f16cad4bdac734717255f29d98e9b6",
                        "test_MT.bam.bai:md5,dbd8f1a5df58115b6f64a9312133d79d"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "test_paired.bam:md5,c6908df90282cc95e846eee31476bb47",
                        "test_paired.bam.bai:md5,77524c609a3001cdca9c7d10dcc14b6a"
                    ]
                ],
                "1": [
                    "versions.yml:md5,0b7a1702e2427e2e30ee7906e41556db",
                    "versions.yml:md5,24321a016768b6449171df86a7f6801b",
                    "versions.yml:md5,3246c6acbc35a76b3ba905ed54ce927b",
                    "versions.yml:md5,bf7294b8c4a1448665b1b771105ffb3c",
                    "versions.yml:md5,bf7294b8c4a1448665b1b771105ffb3c",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399"
                ],
                "bam": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "test_MT.bam:md5,33f16cad4bdac734717255f29d98e9b6",
                        "test_MT.bam.bai:md5,dbd8f1a5df58115b6f64a9312133d79d"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "test_paired.bam:md5,c6908df90282cc95e846eee31476bb47",
                        "test_paired.bam.bai:md5,77524c609a3001cdca9c7d10dcc14b6a"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0b7a1702e2427e2e30ee7906e41556db",
                    "versions.yml:md5,24321a016768b6449171df86a7f6801b",
                    "versions.yml:md5,3246c6acbc35a76b3ba905ed54ce927b",
                    "versions.yml:md5,bf7294b8c4a1448665b1b771105ffb3c",
                    "versions.yml:md5,bf7294b8c4a1448665b1b771105ffb3c",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-06-18T16:46:45.480511171"
    },
    "Rename: bam no chr + fasta chr": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "test_MT.bam:md5,07e733c5bdb64f2f48ceb1eafbb0d120",
                        "test_MT.bam.bai:md5,33b4ed892348364beceb9a2bfba07448"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "test_paired.bam:md5,a4d6b2b5a077ec1d7bdd459c15fe49d0",
                        "test_paired.bam.bai:md5,592b3771c867dfb0b31b14e4a578e795"
                    ]
                ],
                "1": [
                    "versions.yml:md5,0b7a1702e2427e2e30ee7906e41556db",
                    "versions.yml:md5,24321a016768b6449171df86a7f6801b",
                    "versions.yml:md5,3246c6acbc35a76b3ba905ed54ce927b",
                    "versions.yml:md5,bf7294b8c4a1448665b1b771105ffb3c",
                    "versions.yml:md5,bf7294b8c4a1448665b1b771105ffb3c",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399"
                ],
                "bam": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "test_MT.bam:md5,07e733c5bdb64f2f48ceb1eafbb0d120",
                        "test_MT.bam.bai:md5,33b4ed892348364beceb9a2bfba07448"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "test_paired.bam:md5,a4d6b2b5a077ec1d7bdd459c15fe49d0",
                        "test_paired.bam.bai:md5,592b3771c867dfb0b31b14e4a578e795"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0b7a1702e2427e2e30ee7906e41556db",
                    "versions.yml:md5,24321a016768b6449171df86a7f6801b",
                    "versions.yml:md5,3246c6acbc35a76b3ba905ed54ce927b",
                    "versions.yml:md5,bf7294b8c4a1448665b1b771105ffb3c",
                    "versions.yml:md5,bf7294b8c4a1448665b1b771105ffb3c",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-06-18T16:46:58.929558661"
    },
    "No rename: bam chr + fasta chr": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test_illumina_mt.bam",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test_illumina_mt.bam.bai"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai"
                    ]
                ],
                "1": [
                    "versions.yml:md5,3246c6acbc35a76b3ba905ed54ce927b",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399"
                ],
                "bam": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test_illumina_mt.bam",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test_illumina_mt.bam.bai"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3246c6acbc35a76b3ba905ed54ce927b",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399",
                    "versions.yml:md5,ed47e1a8adc5174f9befd7fc57b72399"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-06-18T16:46:33.764782734"
    }
}